crates/compiler/mono/src/layout.rs

```diff
@@ -38,11 +38,397 @@ roc_error_macros::assert_sizeof_default!(Layout, 4 * 8);
 roc_error_macros::assert_sizeof_default!(UnionLayout, 3 * 8);
 roc_error_macros::assert_sizeof_default!(LambdaSet, 3 * 8);
 
+type LayoutResult<'a> = Result<Layout<'a>, LayoutProblem>;
+type RawFunctionLayoutResult<'a> = Result<RawFunctionLayout<'a>, LayoutProblem>;
+
+#[derive(Debug, Clone, Copy)]
+struct CacheMeta {
+    /// Does this cache entry include a recursive structure? If so, what's the recursion variable
+    /// of that structure?
+    has_recursive_structure: Option<Variable>,
+}
+
+impl CacheMeta {
+    #[inline(always)]
+    fn to_criteria(self) -> CacheCriteria {
+        let CacheMeta {
+            has_recursive_structure,
+        } = self;
+        CacheCriteria {
+            has_naked_recursion_pointer: false,
+            has_recursive_structure,
+        }
+    }
+}
+
+/// A single layer of the layout cache.
+/// Snapshots are implemented by operating on new layers, and rollbacks by dropping the latest
+/// layer.
+#[derive(Debug)]
+struct CacheLayer<Result>(MutMap<Variable, (Result, CacheMeta)>);
+
+impl<Result> Default for CacheLayer<Result> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
```

Comment:
I wonder if/how we could improve this representation. How many variables would be in here typically? is it very sparse (can we somehow quantify how sparse)

Reply:
also `FnvHashMap` would probably be a quick win

Reply:
Yeah, those would be good things to measure. I'm not sure we have a correct picture right now, especially before trying to make StorageSubs more dense. But the `statistics` method here should help shed some light. I don't want to try to optimize all at once in this PR, though

Reply:
I switched to using FnvHashMap. Sorry, I had forgotten your suggestion to do that!